File: src/types/audit.ts
/**
 * Nebula Audit Log Schema v1
 *
 * Immutable event model for complete traceability of all actions.
 * Every run produces a complete audit trail: agent decisions, tool calls,
 * code changes, tests, deployments, and documentation writes.
 */

export interface AuditEvent {
  /** Unique event ID */
  id: string;

  /** ISO 8601 timestamp */
  timestamp: string;

  /** The run this event belongs to */
  runId: string;

  /** The workstream this event belongs to */
  workstreamId: string;

  /** The project this event belongs to */
  projectId: string;

  /** Event type */
  type: AuditEventType;

  /** The actor (user or agent) */
  actor: AuditActor;

  /** Event-specific payload */
  payload: AuditPayload;

  /** Parent event ID (for hierarchical tracing) */
  parentEventId?: string;

  /** Span ID for OpenTelemetry correlation */
  spanId?: string;

  /** Trace ID for OpenTelemetry correlation */
  traceId?: string;
}

export type AuditEventType =
  | "run.started"
  | "run.completed"
  | "run.failed"
  | "agent.decision"
  | "agent.delegation"
  | "agent.handoff"
  | "tool.call"
  | "tool.result"
  | "code.read"
  | "code.write"
  | "code.commit"
  | "code.branch.create"
  | "code.merge"
  | "code.rollback"
  | "test.started"
  | "test.passed"
  | "test.failed"
  | "test.report"
  | "gate.evaluated"
  | "gate.passed"
  | "gate.failed"
  | "deploy.started"
  | "deploy.progressed"
  | "deploy.completed"
  | "deploy.rolled-back"
  | "deploy.failed"
  | "documentation.read"
  | "documentation.write"
  | "documentation.consolidated"
  | "linear.issue.created"
  | "linear.issue.updated"
  | "figma.design.ingested"
  | "policy.evaluated"
  | "policy.violation"
  | "user.request"
  | "user.approval"
  | "user.rejection";

export interface AuditActor {
  type: "user" | "agent";
  id: string;
  role?: string;
  name: string;
}

export type AuditPayload =
  | RunPayload
  | AgentDecisionPayload
  | AgentDelegationPayload
  | ToolCallPayload
  | ToolResultPayload
  | CodeChangePayload
  | TestPayload
  | GatePayload
  | DeployPayload
  | DocumentationPayload
  | LinearPayload
  | FigmaPayload
  | PolicyPayload
  | UserRequestPayload;

export interface RunPayload {
  kind: "run";
  status: "started" | "completed" | "failed";
  input?: string;
  output?: string;
  error?: string;
  duration?: number;
}

export interface AgentDecisionPayload {
  kind: "agent.decision";
  decision: string;
  reasoning: string;
  context: string[];
  alternatives: string[];
}

export interface AgentDelegationPayload {
  kind: "agent.delegation";
  fromAgent: string;
  toAgent: string;
  taskDescription: string;
  constraints: string[];
}

export interface ToolCallPayload {
  kind: "tool.call";
  toolId: string;
  operation: string;
  parameters: Record<string, unknown>;
}

export interface ToolResultPayload {
  kind: "tool.result";
  toolId: string;
  operation: string;
  success: boolean;
  result?: unknown;
  error?: string;
}

export interface CodeChangePayload {
  kind: "code.change";
  action: "read" | "write" | "commit" | "branch" | "merge" | "rollback";
  repository: string;
  branch?: string;
  files: FileChange[];
  commitHash?: string;
  message?: string;
}

export interface FileChange {
  path: string;
  action: "added" | "modified" | "deleted" | "renamed";
  additions?: number;
  deletions?: number;
  oldPath?: string;
}

export interface TestPayload {
  kind: "test";
  action: "started" | "passed" | "failed" | "report";
  testType: "unit" | "integration" | "e2e" | "performance" | "security";
  suiteName: string;
  totalTests?: number;
  passed?: number;
  failed?: number;
  skipped?: number;
  duration?: number;
  coveragePercent?: number;
  failures?: TestFailure[];
}

export interface TestFailure {
  testName: string;
  message: string;
  stack?: string;
}

export interface GatePayload {
  kind: "gate";
  gateId: string;
  gateName: string;
  gateType: string;
  result: "passed" | "failed";
  details: string;
}

export interface DeployPayload {
  kind: "deploy";
  action: "started" | "progressed" | "completed" | "rolled-back" | "failed";
  environment: "preview" | "staging" | "production";
  cloud: "azure" | "aws";
  strategy: "canary" | "blue-green" | "rolling" | "slot-swap";
  version: string;
  trafficPercent?: number;
  healthStatus?: string;
  rollbackReason?: string;
}

export interface DocumentationPayload {
  kind: "documentation";
  action: "read" | "write" | "consolidated";
  level: 0 | 1 | 2;
  notePath: string;
  service?: string;
  changesSummary?: string;
}

export interface LinearPayload {
  kind: "linear";
  action: "created" | "updated";
  issueId: string;
  issueTitle: string;
  status?: string;
  url?: string;
}

export interface FigmaPayload {
  kind: "figma";
  action: "ingested";
  fileId: string;
  fileName: string;
  components: string[];
}

export interface PolicyPayload {
  kind: "policy";
  action: "evaluated" | "violation";
  policyRule: string;
  result: "allowed" | "denied";
  reason: string;
}

export interface UserRequestPayload {
  kind: "user.request";
  action: "request" | "approval" | "rejection";
  content: string;
  attachments?: string[];
}

/**
 * Run record — aggregated view of an entire agent run
 */
export interface RunRecord {
  id: string;
  projectId: string;
  workstreamId: string;
  startedAt: string;
  completedAt?: string;
  status: "running" | "completed" | "failed";
  userRequest: string;
  events: AuditEvent[];
  summary?: RunSummary;
}

export interface RunSummary {
  totalEvents: number;
  agentDecisions: number;
  toolCalls: number;
  codeChanges: number;
  testsRun: number;
  testsPassed: number;
  testsFailed: number;
  gatesPassed: number;
  gatesFailed: number;
  deploymentsCompleted: number;
  documentationUpdates: number;
  linearUpdates: number;
  duration: number;
}

/**
 * Create a new audit event
 */
export function createAuditEvent(
  params: Omit<AuditEvent, "id" | "timestamp">
): AuditEvent {
  return {
    id: crypto.randomUUID(),
    timestamp: new Date().toISOString(),
    ...params,
  };
}

/**
 * Create a new run record
 */
export function createRunRecord(
  projectId: string,
  workstreamId: string,
  userRequest: string
): RunRecord {
  return {
    id: crypto.randomUUID(),
    projectId,
    workstreamId,
    startedAt: new Date().toISOString(),
    status: "running",
    userRequest,
    events: [],
  };
}

File: src-tauri/src/audit.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::Mutex;
use uuid::Uuid;
use chrono::Utc;

/// Immutable audit log store.
/// Records every action from user requests to agent decisions, tool calls,
/// code changes, tests, deployments, and documentation writes.
pub struct AuditStore {
    events: Mutex<Vec<AuditEvent>>,
    runs: Mutex<HashMap<String, RunRecord>>,
}

impl AuditStore {
    pub fn new() -> Self {
        Self {
            events: Mutex::new(Vec::new()),
            runs: Mutex::new(HashMap::new()),
        }
    }

    /// Append an event to the immutable log
    pub fn record_event(&self, event: AuditEvent) {
        let mut events = self.events.lock().unwrap();

        // Also add to run record if run exists
        if let Ok(mut runs) = self.runs.lock() {
            if let Some(run) = runs.get_mut(&event.run_id) {
                run.events.push(event.clone());
            }
        }

        events.push(event);
    }

    /// Create a new run record
    pub fn create_run(
        &self,
        project_id: &str,
        workstream_id: &str,
        user_request: &str,
    ) -> String {
        let run_id = Uuid::new_v4().to_string();
        let run = RunRecord {
            id: run_id.clone(),
            project_id: project_id.to_string(),
            workstream_id: workstream_id.to_string(),
            started_at: Utc::now().to_rfc3339(),
            completed_at: None,
            status: "running".to_string(),
            user_request: user_request.to_string(),
            events: Vec::new(),
            summary: None,
        };

        let mut runs = self.runs.lock().unwrap();
        runs.insert(run_id.clone(), run);

        // Record the run start event
        let event = AuditEvent {
            id: Uuid::new_v4().to_string(),
            timestamp: Utc::now().to_rfc3339(),
            run_id: run_id.clone(),
            workstream_id: workstream_id.to_string(),
            project_id: project_id.to_string(),
            event_type: "run.started".to_string(),
            actor: AuditActor {
                actor_type: "user".to_string(),
                id: "system".to_string(),
                role: None,
                name: "User".to_string(),
            },
            payload: serde_json::json!({
                "kind": "run",
                "status": "started",
                "input": user_request
            }),
            parent_event_id: None,
            span_id: None,
            trace_id: None,
        };

        let mut events = self.events.lock().unwrap();
        events.push(event);

        run_id
    }

    /// Complete a run
    pub fn complete_run(&self, run_id: &str, status: &str) {
        let mut runs = self.runs.lock().unwrap();
        if let Some(run) = runs.get_mut(run_id) {
            run.completed_at = Some(Utc::now().to_rfc3339());
            run.status = status.to_string();
            run.summary = Some(Self::compute_summary(&run.events));
        }
    }

    /// Get a run record
    pub fn get_run(&self, run_id: &str) -> Option<RunRecord> {
        let runs = self.runs.lock().unwrap();
        runs.get(run_id).cloned()
    }

    /// Get all events for a project
    pub fn get_events(&self, project_id: &str, limit: usize) -> Vec<AuditEvent> {
        let events = self.events.lock().unwrap();
        events
            .iter()
            .filter(|e| e.project_id == project_id)
            .rev()
            .take(limit)
            .cloned()
            .collect()
    }

    fn compute_summary(events: &[AuditEvent]) -> RunSummary {
        let mut summary = RunSummary {
            total_events: events.len() as u32,
            agent_decisions: 0,
            tool_calls: 0,
            code_changes: 0,
            tests_run: 0,
            tests_passed: 0,
            tests_failed: 0,
            gates_passed: 0,
            gates_failed: 0,
            deployments_completed: 0,
            documentation_updates: 0,
            linear_updates: 0,
            duration: 0,
        };

        for event in events {
            match event.event_type.as_str() {
                "agent.decision" => summary.agent_decisions += 1,
                "tool.call" => summary.tool_calls += 1,
                "code.write" | "code.commit" => summary.code_changes += 1,
                "test.started" => summary.tests_run += 1,
                "test.passed" => summary.tests_passed += 1,
                "test.failed" => summary.tests_failed += 1,
                "gate.passed" => summary.gates_passed += 1,
                "gate.failed" => summary.gates_failed += 1,
                "deploy.completed" => summary.deployments_completed += 1,
                "documentation.write" => summary.documentation_updates += 1,
                "linear.issue.created" | "linear.issue.updated" => summary.linear_updates += 1,
                _ => {}
            }
        }

        summary
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AuditEvent {
    pub id: String,
    pub timestamp: String,
    pub run_id: String,
    pub workstream_id: String,
    pub project_id: String,
    pub event_type: String,
    pub actor: AuditActor,
    pub payload: serde_json::Value,
    pub parent_event_id: Option<String>,
    pub span_id: Option<String>,
    pub trace_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AuditActor {
    pub actor_type: String,
    pub id: String,
    pub role: Option<String>,
    pub name: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RunRecord {
    pub id: String,
    pub project_id: String,
    pub workstream_id: String,
    pub started_at: String,
    pub completed_at: Option<String>,
    pub status: String,
    pub user_request: String,
    pub events: Vec<AuditEvent>,
    pub summary: Option<RunSummary>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RunSummary {
    pub total_events: u32,
    pub agent_decisions: u32,
    pub tool_calls: u32,
    pub code_changes: u32,
    pub tests_run: u32,
    pub tests_passed: u32,
    pub tests_failed: u32,
    pub gates_passed: u32,
    pub gates_failed: u32,
    pub deployments_completed: u32,
    pub documentation_updates: u32,
    pub linear_updates: u32,
    pub duration: u64,
}
